cabal-install/Distribution/Client/ProjectPlanOutput.hs

```diff
@@ -141,6 +141,12 @@ encodePlanAsJson distDirLayout elaboratedInstallPlan elaboratedSharedConfig =
         , "style"      J..= J.String (style2str (elabLocalToProject elab) (elabBuildStyle elab))
         , "pkg-src"    J..= packageLocationToJ (elabPkgSourceLocation elab)
         ] ++
+        [ "pkg-revision" J..=  J.String x
+        | Just x <- [lookup "x-revision" (PD.customFieldsPD (elabPkgDescription elab))]
```

Comment:
strong :-1: from me, sorry!

I'm strongly against adding the `x-revision` field to the `plan.json` file to infer the version; it's a completely unreliable field that is only used internally by Hackage (but other package repositories do not -- the x-revision field would be totally bogus there) for bookkeeping purposes and as a failsafe; any tooling which relies on it to be accurate is plain and simple broken!

Also, the revision number is a concept that breaks down completely for e.g. Hackage overlays; the only way to accurately identify a revision is via its SHA256. I plan to add support to cabal for pinning to revisions but I do not intend to support pinning by revision number; only by either SHA256 or by timestamp (the latter addressing scheme will likely be the first to be implemented as it's the one that's close to the `--index-state` design).

Reply:
That indeed sounds like a saner approach to me.   Let me just state that I would very much be interested for the `plan.json` to contain information about what revision is being used.    The approach I implemented might be too fragile, so let's not merge it, but I would very much a less fragile approach to be available.